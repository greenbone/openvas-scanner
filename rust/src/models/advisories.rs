// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

use std::collections::HashMap;

/// Represents an advisory json file for notus product.
#[cfg_attr(feature = "serde_support", derive(serde::Deserialize))]
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ProductsAdvisories {
    /// Version of the advisory file
    pub version: String,
    /// SPDX license identifier
    #[cfg_attr(feature = "serde_support", serde(rename = "spdx-license-identifier"))]
    pub license_identifier: String,
    /// Copyright
    pub copyright: String,
    /// Vulnerability Family
    pub family: String,
    /// List of Advisories
    #[cfg_attr(feature = "serde_support", serde(default))]
    pub advisories: Vec<Advisory>,
}

/// Represents an advisory json file for notus product.
#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize)
)]
pub struct Advisory {
    /// The advisory's title.
    pub title: String,
    /// The advisory's ID.
    pub oid: String,
    /// Creation Date
    pub creation_date: u64,
    /// Last modification date
    pub last_modification: u64,
    /// Advisory ID
    pub advisory_id: String,
    /// Advisory xref
    pub advisory_xref: String,
    /// Advisory contains a CVE that is listed in the catalog of Known Exploited CVEs from CISA
    #[cfg_attr(feature = "serde_support", serde(default))]
    pub cisa_kev: bool,
    /// List of cves
    #[cfg_attr(feature = "serde_support", serde(default))]
    pub cves: Vec<String>,
    /// Summary
    pub summary: String,
    /// Insight
    #[cfg_attr(feature = "serde_support", serde(default))]
    pub insight: String,
    /// Affected
    pub affected: String,
    /// List of xrefs
    #[cfg_attr(feature = "serde_support", serde(default))]
    pub xrefs: Vec<String>,
    /// Quality of detection
    pub qod_type: String,
    /// Severity
    pub severity: Severity,
}

/// A single vulnerability from an advisory file to be stored
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize)
)]
#[derive(Default, Debug, Clone, PartialEq, Eq)]
pub struct Vulnerability {
    /// VT Parameters
    pub vt_params: Vec<String>,
    /// Creation Date
    pub creation_date: u64,
    /// Last modification date
    pub last_modification: u64,
    /// Summary
    pub summary: String,
    /// Impact
    pub impact: String,
    /// Affected
    pub affected: String,
    /// Insight
    pub insight: String,
    /// Solution
    pub solution: String,
    /// Solution Type
    pub solution_type: String,
    /// Vuldetect
    pub vuldetect: String,
    /// Quality of detection
    pub qod_type: String,
    /// Severity vector
    pub severity_vector: String,
    /// File name
    pub filename: String,
    /// All references: xrefs, cves, xrefs, advisory xrefs and advisory id.
    pub refs: HashMap<String, Vec<String>>,
    /// Vulnerability Family
    pub family: String,
    /// Title
    pub name: String,
    /// Category
    pub category: String,
}

/// Severity
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize)
)]
#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
pub struct Severity {
    /// Origin of the severity
    pub origin: String,
    /// severity date
    pub date: u64,
    /// Cvss version v2
    #[cfg_attr(
        feature = "serde_support",
        serde(skip_serializing_if = "Option::is_none")
    )]
    pub cvss_v2: Option<String>,
    /// cvss vector v3
    #[cfg_attr(
        feature = "serde_support",
        serde(skip_serializing_if = "Option::is_none")
    )]
    pub cvss_v3: Option<String>,
}

pub struct ProductsAdvisoriesIterator<'a> {
    products_advisories: &'a ProductsAdvisories,
    index: usize,
}

impl<'a> Iterator for ProductsAdvisoriesIterator<'a> {
    type Item = &'a Advisory;

    fn next(&mut self) -> Option<&'a Advisory> {
        if self.index < self.products_advisories.advisories.len() {
            let result = Some(&self.products_advisories.advisories[self.index]);
            self.index += 1;
            result
        } else {
            None
        }
    }
}

impl ProductsAdvisories {
    pub fn iter(&self) -> ProductsAdvisoriesIterator {
        ProductsAdvisoriesIterator {
            products_advisories: self,
            index: 0,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct VulnerabilityData {
    pub adv: Advisory,
    pub family: String,
    pub filename: String,
}

impl From<VulnerabilityData> for Vulnerability {
    fn from(data: VulnerabilityData) -> Self {
        let sv = match data.adv.severity.cvss_v2 {
            Some(cvss) => cvss,
            None => match data.adv.severity.cvss_v3 {
                Some(cvss) => cvss,
                None => "".to_string(),
            },
        };

        let mut refs = HashMap::new();
        let mut url = data.adv.xrefs.clone();
        url.push(data.adv.advisory_xref.clone());
        if data.adv.cisa_kev {
            refs.insert(
                "CISA".to_string(),
                vec!["Known Exploited Vulnerability (KEV) catalog".to_string()],
            );
            url.push("https://www.cisa.gov/known-exploited-vulnerabilities-catalog".to_string());
        }
        refs.insert("URL".to_string(), url);

        Self {
            vt_params: Vec::new(),
            creation_date: data.adv.creation_date,
            last_modification: data.adv.last_modification,
            summary: data.adv.summary,
            impact: "".to_string(),
            affected: data.adv.affected,
            insight: data.adv.insight,
            solution: "Please install the updated package(s).".to_string(),
            solution_type: "VendorFix".to_string(),
            vuldetect: "Checks if a vulnerable package version is present on the target host."
                .to_string(),
            qod_type: data.adv.qod_type,
            severity_vector: sv,
            filename: data.filename,
            refs,
            family: data.family,
            name: data.adv.title,
            category: "3".to_string(),
        }
    }
}
