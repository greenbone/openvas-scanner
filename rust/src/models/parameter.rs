// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

use std::fmt::Display;

#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
#[cfg_attr(
    feature = "serde_support",
    derive(serde::Serialize, serde::Deserialize)
)]
/// Represents a parameter for a VTS configuration.
pub struct Parameter {
    /// The ID of the parameter.
    pub id: u16,
    /// The value of the parameter.
    pub value: String,
}

impl Display for Parameter {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}: {}", self.id, self.value)
    }
}
