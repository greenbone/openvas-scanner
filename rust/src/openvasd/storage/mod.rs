// SPDX-FileCopyrightText: 2024 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

pub mod file;
pub mod inmemory;
pub mod redis;
pub mod results;

use std::{
    collections::HashMap,
    path::{Path, PathBuf},
    sync::Arc,
};

use async_trait::async_trait;
use scannerlib::{
    models::{self, scanner::ScanResults, FeedType, Scan, Status, VulnerabilityData},
    storage::{error::StorageError, inmemory::InMemoryStorage, items::nvt::Nvt, Dispatcher},
};

use crate::{config::Config, controller::ClientHash, crypt};
use scannerlib::{
    feed::{self, HashSumNameLoader, Update},
    nasl::FSPluginLoader,
    notus::{AdvisoryLoader, HashsumAdvisoryLoader},
};

#[derive(Debug)]
pub enum Error {
    Serialization,
    NotFound,
    Storage(Box<dyn std::error::Error + Send + Sync>),
}

impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use Error::*;
        match self {
            NotFound => write!(f, "not found"),
            Serialization => write!(f, "serialization error"),
            Storage(e) => write!(f, "storage error: {e}"),
        }
    }
}

impl std::error::Error for Error {}

impl From<serde_json::Error> for Error {
    fn from(_: serde_json::Error) -> Self {
        Self::Serialization
    }
}

impl From<crypt::ParseError> for Error {
    fn from(_: crypt::ParseError) -> Self {
        Self::Serialization
    }
}
impl From<std::string::FromUtf8Error> for Error {
    fn from(_: std::string::FromUtf8Error) -> Self {
        Self::Serialization
    }
}

impl From<crate::storage::Error> for models::scanner::Error {
    fn from(value: crate::storage::Error) -> Self {
        Self::Unexpected(format!("{value:?}"))
    }
}

impl From<scannerlib::storage::infisto::Error> for Error {
    fn from(e: scannerlib::storage::infisto::Error) -> Self {
        Self::Storage(Box::new(e))
    }
}

impl From<scannerlib::storage::redis::DbError> for Error {
    fn from(value: scannerlib::storage::redis::DbError) -> Self {
        Error::Storage(Box::new(value))
    }
}

impl From<StorageError> for Error {
    fn from(value: StorageError) -> Self {
        Error::Storage(Box::new(value))
    }
}

#[async_trait]
pub trait ScanIDClientMapper {
    async fn add_scan_client_id(&self, scan_id: String, client_id: ClientHash)
    -> Result<(), Error>;
    async fn remove_scan_id<I>(&self, scan_id: I) -> Result<(), Error>
    where
        I: AsRef<str> + Send + 'static;

    async fn get_scans_of_client_id(&self, client_id: &ClientHash) -> Result<Vec<String>, Error>;

    async fn is_client_allowed<I>(&self, scan_id: I, client_id: &ClientHash) -> Result<bool, Error>
    where
        I: AsRef<str> + Send + 'static,
    {
        let scans = self.get_scans_of_client_id(client_id).await?;
        let sid = scan_id.as_ref();
        for id in scans {
            if id == sid {
                return Ok(true);
            }
        }
        Ok(false)
    }
}

#[async_trait]
/// A trait for getting the progress of a scan, the scan itself with decrypted credentials and
/// encrypted as well as results.
///
/// The main usage of this trait is in the controller and when transforming a scan to a osp
pub trait ProgressGetter {
    /// Returns the scan.
    async fn get_scan(&self, id: &str) -> Result<(Scan, Status), Error>;
    /// Returns the scan with decrypted passwords.
    ///
    /// This method should only be used when the password is required. E.g.
    /// when transforming a scan to a osp command.
    async fn get_decrypted_scan(&self, id: &str) -> Result<(Scan, Status), Error>;
    /// Returns all scans.
    async fn get_scan_ids(&self) -> Result<Vec<String>, Error>;
    /// Returns the status of a scan.
    async fn get_status(&self, id: &str) -> Result<Status, Error>;
    /// Returns the results of a scan as json bytes.
    ///
    /// OpenVASD just stores to results without processing them therefore we
    /// can just return the json bytes.
    async fn get_results(
        &self,
        id: &str,
        from: Option<usize>,
        to: Option<usize>,
    ) -> Result<Box<dyn Iterator<Item = Vec<u8>> + Send>, Error>;
}

pub type Hash = String;

#[derive(Clone, Debug, PartialEq, Eq)]
/// Contains the hash values of the sha256sums for specific feeds
pub struct FeedHash {
    pub hash: Hash,
    pub path: PathBuf,
    pub typus: FeedType,
}

impl FeedHash {
    pub fn advisories<S>(p: S) -> Self
    where
        S: AsRef<Path>,
    {
        FeedHash {
            hash: String::new(),
            path: p.as_ref().to_path_buf(),
            typus: FeedType::Advisories,
        }
    }

    pub fn nasl<S>(p: S) -> Self
    where
        S: AsRef<Path>,
    {
        FeedHash {
            hash: String::new(),
            path: p.as_ref().to_path_buf(),
            typus: FeedType::NASL,
        }
    }
}

#[async_trait]
/// Handles NVT specifics.
///
/// Usually it parses nasl feed and notus feed to generate and store nvts.
pub trait NVTStorer {
    /// Synchronizes feed based on the given hash.
    ///
    /// This method is called when the sha256sums is changed. It will then go through the feed
    /// directories and update the meta information.
    async fn synchronize_feeds(&self, hash: Vec<FeedHash>) -> Result<(), Error>;

    /// Retrieves just all oids.
    async fn oids(&self) -> Result<Vec<String>, Error>;

    /// Retrieves NVTs.
    async fn vts<'a>(&self) -> Result<Vec<Nvt>, Error>;

    /// Retrieves a NVT.
    ///
    async fn vt_by_oid(&self, oid: &str) -> Result<Option<Nvt>, Error>;

    /// Returns the currently stored feed hash.
    async fn feed_hash(&self) -> Vec<FeedHash>;

    /// Returns the current feed version, if any.
    async fn current_feed_version(&self) -> Result<String, Error>;
}

#[async_trait]
/// A trait for storing scans.
///
/// The main usage of this trait is in the controller and when a user inserts or removes a scan.
pub trait ScanStorer {
    /// Inserts a scan.
    async fn insert_scan(&self, t: Scan) -> Result<(), Error>;
    /// Removes a scan.
    async fn remove_scan(&self, id: &str) -> Result<(), Error>;
    /// Updates a status of a scan.
    ///
    /// This is required when a scan is started or stopped.
    async fn update_status(&self, id: &str, status: Status) -> Result<(), Error>;
}

#[async_trait]
/// A trait for appending results from a different source.
///
/// This is used when a scan is started and the results are fetched from ospd.
pub trait AppendFetchResult {
    async fn append_fetched_result(&self, results: Vec<ScanResults>) -> Result<(), Error>;
}
#[async_trait]
impl<T> AppendFetchResult for Arc<T>
where
    T: AppendFetchResult + Send + Sync,
{
    async fn append_fetched_result(&self, results: Vec<ScanResults>) -> Result<(), Error> {
        self.as_ref().append_fetched_result(results).await
    }
}

#[async_trait]
impl<T> ProgressGetter for Arc<T>
where
    T: ProgressGetter + Send + Sync,
{
    async fn get_scan(&self, id: &str) -> Result<(Scan, Status), Error> {
        self.as_ref().get_scan(id).await
    }

    async fn get_decrypted_scan(&self, id: &str) -> Result<(Scan, Status), Error> {
        self.as_ref().get_decrypted_scan(id).await
    }

    async fn get_scan_ids(&self) -> Result<Vec<String>, Error> {
        self.as_ref().get_scan_ids().await
    }

    async fn get_status(&self, id: &str) -> Result<Status, Error> {
        self.as_ref().get_status(id).await
    }

    async fn get_results(
        &self,
        id: &str,
        from: Option<usize>,
        to: Option<usize>,
    ) -> Result<Box<dyn Iterator<Item = Vec<u8>> + Send>, Error> {
        self.as_ref().get_results(id, from, to).await
    }
}

#[async_trait]
impl<T> ScanStorer for Arc<T>
where
    T: ScanStorer + Send + Sync,
{
    async fn insert_scan(&self, t: Scan) -> Result<(), Error> {
        self.as_ref().insert_scan(t).await
    }
    async fn remove_scan(&self, id: &str) -> Result<(), Error> {
        self.as_ref().remove_scan(id).await
    }

    async fn update_status(&self, id: &str, status: Status) -> Result<(), Error> {
        self.as_ref().update_status(id, status).await
    }
}

#[async_trait]
impl<T> NVTStorer for Arc<T>
where
    T: NVTStorer + Send + Sync,
{
    async fn synchronize_feeds(&self, hash: Vec<FeedHash>) -> Result<(), Error> {
        self.as_ref().synchronize_feeds(hash).await
    }

    async fn oids(&self) -> Result<Vec<String>, Error> {
        self.as_ref().oids().await
    }

    async fn vts<'a>(&self) -> Result<Vec<Nvt>, Error> {
        self.as_ref().vts().await
    }

    async fn vt_by_oid(&self, oid: &str) -> Result<Option<Nvt>, Error> {
        self.as_ref().vt_by_oid(oid).await
    }

    async fn feed_hash(&self) -> Vec<FeedHash> {
        self.as_ref().feed_hash().await
    }

    async fn current_feed_version(&self) -> Result<String, Error> {
        self.as_ref().current_feed_version().await
    }
}

#[async_trait]
impl<T> ScanIDClientMapper for Arc<T>
where
    T: ScanIDClientMapper + Send + Sync,
{
    async fn add_scan_client_id(
        &self,
        scan_id: String,
        client_id: ClientHash,
    ) -> Result<(), Error> {
        self.as_ref().add_scan_client_id(scan_id, client_id).await
    }
    async fn remove_scan_id<I>(&self, scan_id: I) -> Result<(), Error>
    where
        I: AsRef<str> + Send + 'static,
    {
        self.as_ref().remove_scan_id(scan_id).await
    }

    async fn get_scans_of_client_id(&self, client_id: &ClientHash) -> Result<Vec<String>, Error> {
        self.as_ref().get_scans_of_client_id(client_id).await
    }

    async fn is_client_allowed<I>(&self, scan_id: I, client_id: &ClientHash) -> Result<bool, Error>
    where
        I: AsRef<str> + Send + 'static,
    {
        self.as_ref().is_client_allowed(scan_id, client_id).await
    }
}

#[async_trait]
/// Combines the traits `ProgressGetter`, `ScanStorer` and `AppendFetchResult`.
pub trait Storage:
    ProgressGetter + ScanStorer + AppendFetchResult + NVTStorer + ScanIDClientMapper
{
}

#[async_trait]
impl<T> Storage for T where
    T: ProgressGetter + ScanStorer + AppendFetchResult + NVTStorer + ScanIDClientMapper
{
}

/// A storage type that can be created from a Config and a list of feeds.
pub trait FromConfigAndFeeds: ResultHandler + Storage + Sized {
    fn from_config_and_feeds(
        config: &Config,
        feeds: Vec<FeedHash>,
    ) -> Result<Self, Box<dyn std::error::Error + Send + Sync>>;
}

async fn update_notus_feed(p: PathBuf, store: Arc<InMemoryStorage>) -> Result<(), Error> {
    let notus_advisories_path = p;
    store.clean_advisories()?;

    tokio::task::spawn_blocking(move || {
        tracing::debug!("starting notus feed update");
        let loader = FSPluginLoader::new(notus_advisories_path);
        let advisories_files = HashsumAdvisoryLoader::new(loader.clone())?;
        for filename in advisories_files.get_advisories()?.iter() {
            let advisories = advisories_files.load_advisory(filename)?;

            for adv in advisories.advisories {
                let data = VulnerabilityData {
                    adv,
                    family: advisories.family.clone(),
                    filename: filename.to_owned(),
                };

                store.dispatch((), data)?;
            }
        }
        tracing::debug!("finished notus feed update");
        Ok(())
    })
    .await
    .expect("notus handler to be executed.")
}

async fn update_nasl_feed(p: PathBuf, store: Arc<InMemoryStorage>) -> Result<(), Error> {
    let nasl_feed_path = p;
    store.as_ref().clean_vts()?;

    tracing::debug!("starting nasl feed update");
    let oversion = "0.1";
    let loader = FSPluginLoader::new(nasl_feed_path);
    let verifier = HashSumNameLoader::sha256(&loader)?;

    let fu = Update::init(oversion, 5, &loader, &store, verifier);
    fu.perform_update().await?;
    tracing::debug!("finished nasl feed update");
    Ok(())
}

pub trait ResultHandler {
    fn underlying_storage(&self) -> &Arc<InMemoryStorage>;
    fn handle_result<E>(&self, key: &str, result: models::Result) -> Result<(), E>
    where
        E: From<StorageError>;
    fn remove_result<E>(&self, key: &str, idx: Option<usize>) -> Result<Vec<models::Result>, E>
    where
        E: From<StorageError>;
}
