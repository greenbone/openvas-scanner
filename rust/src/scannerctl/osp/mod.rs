// SPDX-FileCopyrightText: 2024 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

use std::io::BufRead;
use std::{io::BufReader, path::PathBuf, sync::Arc};

use clap::{arg, value_parser, Arg, ArgAction, Command};
use scannerlib::models::{self, Parameter, Scan, VT};
use scannerlib::storage::{self, DefaultDispatcher, StorageError};
use start_scan::{StartScan, VtSelection};

use crate::{CliError, CliErrorKind};
use scannerlib::storage::item::{NVTField, NVTKey};
use scannerlib::storage::Field;
use scannerlib::storage::Retrieve;
mod start_scan;

pub fn extend_args(cmd: Command) -> Command {
    cmd.subcommand(crate::add_verbose(
        Command::new("osp")
            .about("Transforms a osp start-scan xml to a scan json for openvasd. ")
            .arg(
                arg!(-p --path <FILE> "Path to the feed.")
                    .required(false)
                    .value_parser(value_parser!(PathBuf)),
            )
            .arg(
                Arg::new("ospd_xml")
                    .required(false),
            )
            .arg(
                arg!(-b --back "Serializes start scan command and pretty prints it back to stdout.")
                    .required(false)
                    .action(ArgAction::SetTrue),
            ),
    ))
}

pub async fn may_transform_start_scan<R, S>(
    print_back: bool,
    feed: Option<S>,
    reader: R,
) -> Option<Result<String, CliErrorKind>>
where
    R: BufRead,
    S: storage::Retriever,
{
    match quick_xml::de::from_reader(reader) {
        Ok(x) if print_back => Some(Ok(format!("{x}"))),
        Ok(x) if feed.is_some() => Some(transform_start_scan(feed.unwrap(), x).await),
        Ok(_) => Some(Err(CliErrorKind::MissingArguments(
            vec!["path".to_string()],
        ))),
        Err(_) => None,
    }
}

async fn transform_vts<S>(feed: S, vts: VtSelection) -> Result<Vec<models::VT>, CliErrorKind>
where
    S: storage::Retriever,
{
    let mut result: Vec<_> = vts
        .vt_single
        .into_iter()
        .flatten()
        .map(|x| VT {
            oid: x.id,
            parameters: x
                .vt_value
                .into_iter()
                .flatten()
                .filter_map(|x| x.id.parse().ok().map(|y| (y, x.text)))
                .filter_map(|(id, x)| x.map(|v| Parameter { id, value: v }))
                .collect(),
        })
        .collect();
    let gvts = vts.vt_group.into_iter().flatten().filter_map(|x| {
        match x.filter.split_once('=').map(|(k, v)| (k.trim(), v.trim())) {
            Some(("family", v)) => Some(v.to_string()),
            filter => {
                tracing::warn!(?filter, "only family is supported, ignoring entry");
                None
            }
        }
    });

    // we iterate here to return an error when storage is behaving in an unexpected fashion
    for family in gvts {
        let fvts: Vec<VT> = match feed.retry_retrieve_by_field(
            Field::NVT(NVTField::Family(family.to_string())),
            Retrieve::NVT(Some(NVTKey::Oid)),
            5,
        ) {
            Ok(x) => x
                .flat_map(|(_, f)| match &f {
                    Field::NVT(NVTField::Oid(oid)) => Some(VT {
                        oid: oid.clone(),
                        ..Default::default()
                    }),
                    _ => None,
                })
                .collect(),
            Err(StorageError::NotFound(_)) => {
                tracing::debug!(family, "not found");
                Vec::new()
            }
            Err(e) => return Err(e.into()),
        };
        result.extend(fvts);
    }
    result.sort();
    Ok(result)
}

async fn transform_start_scan<S>(feed: S, sc: StartScan) -> Result<String, CliErrorKind>
where
    S: storage::Retriever,
{
    // currently we ignore the previous order as the scanner will reorder
    // when scheduling internally anyway.
    let scan = Scan {
        scan_id: sc.id.unwrap_or_default(),
        scan_preferences: sc.scanner_params.values,
        target: sc.targets.target.into(),
        vts: transform_vts(feed, sc.vt_selection).await?,
    };
    let scan_json = match serde_json::to_string_pretty(&scan) {
        Ok(s) => s,
        Err(e) => return Err(e.into()),
    };
    Ok(scan_json)
}

pub async fn run(root: &clap::ArgMatches) -> Option<Result<(), CliError>> {
    let (args, _) = crate::get_args_set_logging(root, "osp")?;

    let feed = match args.get_one::<PathBuf>("path") {
        Some(feed) => {
            tracing::info!("loading feed. This may take a while.");
            let storage = Arc::new(DefaultDispatcher::new());
            crate::feed::update::run(Arc::clone(&storage), feed.to_owned(), false)
                .await
                .unwrap();
            tracing::info!("feed loaded.");
            Some(storage)
        }
        None => None,
    };

    let config = args.get_one::<String>("ospd_xml");
    let mut bufreader: BufReader<Box<dyn std::io::Read>> = {
        if let Some(config) = config {
            let file = match std::fs::File::open(config) {
                Ok(x) => x,
                Err(e) => return Some(Err(e.into())),
            };
            BufReader::new(Box::new(file))
        } else {
            BufReader::new(Box::new(std::io::stdin()))
        }
    };
    let print_back = args.get_one::<bool>("back").cloned().unwrap_or_default();
    // currently we just support start scan if that changes chain the options.
    let output = may_transform_start_scan(print_back, feed, &mut bufreader).await;
    let result = match output {
        Some(Ok(x)) => {
            println!("{x}");
            Ok(())
        }
        Some(Err(e)) => Err(CliError {
            filename: config.cloned().unwrap_or_default(),
            kind: e,
        }),
        None => Err(CliError {
            filename: config.cloned().unwrap_or_default(),
            kind: CliErrorKind::Corrupt("Unknown ospd command.".to_string()),
        }),
    };

    Some(result)
}

#[cfg(test)]
mod tests {
    use std::io::Cursor;

    use scannerlib::storage::{item::NVTField, ContextKey, DefaultDispatcher, Field};
    use storage::Dispatcher;

    use super::*;

    #[tokio::test]
    #[tracing_test::traced_test]
    async fn print_scan_json() {
        let input = r#"
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<start_scan parallel="20" scan_id="36389b56-f5a0-11e9-bba4-482ae354ac4c">
    <targets>
        <target>
            <hosts>127.0.0.1</hosts>
            <ports>T:80-80,443-443</ports>
            <alive_test_ports>T:80-80,443-443</alive_test_ports>
            <alive_test>2</alive_test>
            <credentials>
                <credential type="up" service="ssh" port="22">
                  <password>PASSWORD</password>
                  <username>USER</username>
                </credential>
              </credentials>
            <exclude_hosts>localhost</exclude_hosts>
        </target>
    </targets>
    <vt_selection>
        <vt_group filter="family=A"/>        
        <vt_single id="1.3.6.1.4.1.25623.1.0.100151">
            <vt_value id="1">postgres</vt_value>
            <vt_value id="2"/>
        </vt_single>
    </vt_selection>
    <scanner_params>
        <use_mac_addr>0</use_mac_addr>
        <checks_read_timeout>5</checks_read_timeout>
        <cgi_path>/cgi-bin:/scripts</cgi_path>
        <time_between_request>0</time_between_request>
        <vhosts_ip/>
        <vhosts/>
    </scanner_params>
</start_scan>
    "#;
        let reader = BufReader::new(Cursor::new(input));
        let d = DefaultDispatcher::new();
        let dispatch = |k: &str, f: &str| {
            let key = ContextKey::FileName(format!("{k}.nasl"));
            d.dispatch(&key, Field::NVT(NVTField::Family(f.into())))
                .unwrap();

            d.dispatch(&key, Field::NVT(NVTField::Oid(k.into())))
                .unwrap();
        };
        dispatch("0", "A");
        dispatch("1", "A");
        dispatch("2", "A");
        dispatch("3", "A");

        let output = may_transform_start_scan(false, Some(d), reader)
            .await
            .unwrap()
            .unwrap();
        insta::assert_snapshot!(output);
    }

    #[tokio::test]
    #[tracing_test::traced_test]
    async fn print_back() {
        let input = r#"
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<start_scan parallel="20" scan_id="36389b56-f5a0-11e9-bba4-482ae354ac4c">
    <targets>
        <target>
            <hosts>127.0.0.1</hosts>
            <ports>T:80-80,443-443</ports>
            <alive_test_ports>T:80-80,443-443</alive_test_ports>
            <alive_test>2</alive_test>
            <credentials>
                <credential type="up" service="ssh" port="22">
                  <password>PASSWORD</password>
                  <username>USER</username>
                </credential>
              </credentials>
            <exclude_hosts>localhost</exclude_hosts>
        </target>
    </targets>
    <vt_selection>
        <vt_group filter="family=AIX Local Security Checks"/>        
        <vt_single id="1.3.6.1.4.1.25623.1.0.100151">
            <vt_value id="1">postgres</vt_value>
            <vt_value id="2"/>
        </vt_single>
    </vt_selection>
    <scanner_params>
        <use_mac_addr>0</use_mac_addr>
        <checks_read_timeout>5</checks_read_timeout>
        <cgi_path>/cgi-bin:/scripts</cgi_path>
        <time_between_request>0</time_between_request>
        <vhosts_ip/>
        <vhosts/>
    </scanner_params>
</start_scan>
    "#;
        let reader = BufReader::new(Cursor::new(input));
        let output = may_transform_start_scan::<_, DefaultDispatcher>(true, None, reader)
            .await
            .unwrap()
            .unwrap();
        insta::assert_snapshot!(output);
    }
}
