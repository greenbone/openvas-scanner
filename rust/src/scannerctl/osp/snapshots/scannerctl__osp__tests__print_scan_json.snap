---
source: src/scannerctl/osp/mod.rs
expression: output
---
{
  "scan_id": "36389b56-f5a0-11e9-bba4-482ae354ac4c",
  "target": {
    "hosts": [
      "127.0.0.1"
    ],
    "ports": [
      {
        "protocol": "tcp",
        "range": [
          {
            "start": 80,
            "end": 80
          }
        ]
      },
      {
        "protocol": "tcp",
        "range": [
          {
            "start": 443,
            "end": 443
          }
        ]
      }
    ],
    "excluded_hosts": [
      "localhost"
    ],
    "credentials": [
      {
        "service": "ssh",
        "port": 22,
        "up": {
          "username": "USER",
          "password": "PASSWORD"
        }
      }
    ],
    "alive_test_ports": [
      {
        "protocol": "tcp",
        "range": [
          {
            "start": 80,
            "end": 80
          }
        ]
      },
      {
        "protocol": "tcp",
        "range": [
          {
            "start": 443,
            "end": 443
          }
        ]
      }
    ],
    "alive_test_methods": [
      "icmp"
    ],
    "reverse_lookup_unify": null,
    "reverse_lookup_only": null
  },
  "scan_preferences": [
    {
      "id": "use_mac_addr",
      "value": "0"
    },
    {
      "id": "checks_read_timeout",
      "value": "5"
    },
    {
      "id": "cgi_path",
      "value": "/cgi-bin:/scripts"
    },
    {
      "id": "time_between_request",
      "value": "0"
    },
    {
      "id": "vhosts_ip",
      "value": ""
    },
    {
      "id": "vhosts",
      "value": ""
    }
  ],
  "vts": [
    {
      "oid": "0",
      "parameters": []
    },
    {
      "oid": "1",
      "parameters": []
    },
    {
      "oid": "1.3.6.1.4.1.25623.1.0.100151",
      "parameters": [
        {
          "id": 1,
          "value": "postgres"
        }
      ]
    },
    {
      "oid": "2",
      "parameters": []
    },
    {
      "oid": "3",
      "parameters": []
    }
  ]
}
